src/Framework/Constraint/Count.php

```diff
@@ -116,4 +81,82 @@ protected function failureDescription($other): string
             $this->expectedCount
         );
     }
+
+    private function getCountOfTraversable(Traversable $traversable): int
+    {
+        if ($this->traversableCounts === null) {
+            $this->traversableCounts = new \SplObjectStorage();
+        }
+
+        while ($traversable instanceof IteratorAggregate) {
+            $traversable = $traversable->getIterator();
+        }
+
+        if ($this->traversableCounts->contains($traversable)) {
+            return $this->traversableCounts[$traversable];
+        }
+
+        if ($traversable instanceof Iterator) {
+            return $this->countIteratorWithoutChangingPositionIfPossible($traversable);
+        }
+
+        return $this->getCountOfTraversableThatIsNotIteratorOrIteratorAggregate($traversable);
+    }
+
+    private function countIteratorWithoutChangingPositionIfPossible(Iterator $iterator): int
+    {
+        $key   = $iterator->key();
+        $count = $this->getCountOfIterator($iterator);
+
+        $this->traversableCounts->attach($iterator, $count);
+
+        if ($this->iteratorIsRewindable($iterator)) {
+            $this->traversableCounts->detach($iterator);
+            $this->rewindIterator($iterator, $key);
+        }
+
+        return $count;
+    }
+
+    private function iteratorIsRewindable(Iterator $iterator): bool
+    {
+        try {
+            $iterator->rewind();
+        } catch (\Exception $e) {
+            return false;
+        }
+
+        return !($iterator instanceof \NoRewindIterator);
```

Comment:
Thoughts:
- Reorder `getCountOfIterator` before this function
- There is no need to try rewind if it is a `NoRewindIterator`
- If you cannot test whether an iterator is rewindable without actually trying to rewind it, maybe should you merge this function with `rewindIterator` below, so that the current position is unchanged (restored)

Reply:
> Reorder getCountOfIterator before this function

Done. That is some hard looking for things that can be improved there! In some way this is odd given this code is much higher quality than a lot of what I looked at while making these changes.

> If you cannot test whether an iterator is rewindable without actually trying to rewind it, maybe should you merge this function with rewindIterator below...

That was what the code was doing at some point. But then you have both the check, the rewind and a `traversableCounts->detach()` in there. Esp the detach is badly placed then.

Reply:
What about something like this?

```php
    private function getCountOfTraversable(Traversable $traversable): int
    {
        ...

        if ($this->traversableCounts->contains($traversable)) {
            return $this->traversableCounts[$traversable];
        }

        if ($traversable instanceof Iterator) {
            $count = $this->countIteratorWithoutChangingPositionIfPossible($traversable);
        } else {
            $count = \iterator_count($traversable);
        }

        $this->traversableCounts->attach($traversable, $count);

        return $count;
    }

    /**
     * Returns the total number of iterations from a iterator.
     * This will fully exhaust the iterator if it is not rewindable.
     */
    private function countIteratorWithoutChangingPositionIfPossible(Iterator $iterator): int
    {
        $key = $iterator->key();

        // Note: iterator_count calls rewind internally.
        try {
            $count = \iterator_count($iterator);
        } catch (\Exception $e) {
            // Fall back to manual iteration.
            for ($count = 0; $iterator->valid(); $iterator->next()) {
                ++$count;
            }

            return $count;
        }

        // Manually rewind $iterator to previous key, since iterator_count
        // moves pointer.
        if ($key !== null) {
            try {
                $iterator->rewind();
            } catch (\Exception $e) {
                return $count;
            }

            while ($iterator->valid() && $key !== $iterator->key()) {
                $iterator->next();
            }
        }

        return $count;
    }
```

It's simpler (I find) and keeps closer to the original code without needing to check for specific implementations like `NoRewindIterator` or even `Generator`. Maybe it can still be simplified or otherwise improved?